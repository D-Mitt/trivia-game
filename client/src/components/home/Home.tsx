import React, { useCallback, useEffect, useState } from "react"
import Button from "react-bootstrap/Button"
import { useDispatch, useSelector } from "react-redux"
import { Navigate } from "react-router-dom"
import { NIL as NIL_UUID } from "uuid"
import { createNewSoloGame, joinMultiplayerGame } from "../../store/game/GameActions"
import { State } from "../../store/store"
import "../../stylesheets/home.css"

// The landing page of the app
const Home = () => {
  // Set up state
  const dispatch = useDispatch()
  const isSearchingForGame = useSelector((state: State) => state.game.isSearchingForGame)
  const isCreatingSoloGame = useSelector((state: State) => state.game.isCreatingSoloGame)
  const gameId = useSelector((state: State) => state.game.gameId)

  // We found a game
  if (gameId !== NIL_UUID) {
    return <Navigate to={`/games/${gameId}`} />
  }
  
  // Component that finds or starts a new multiplayer game
  const FindGameButton = ({classes}: any) => {
    const [requestJoinMultiplayerGame, setRequestJoinMultiplayerGame] = useState(false)
    const handleClick = useCallback(() => {
      setRequestJoinMultiplayerGame(true)
    }, [requestJoinMultiplayerGame])

    useEffect(() => {
      if (requestJoinMultiplayerGame) {
        dispatch(joinMultiplayerGame())
      }
    }, [requestJoinMultiplayerGame])

    return (
      <Button
        className={`${classes} mt-5`}
        variant="success"
        size="lg"
        onClick={!isSearchingForGame ? () => {handleClick()} : undefined}
        disabled={isSearchingForGame || isCreatingSoloGame}
      >
        {isSearchingForGame ? "Searching..." : "Find a Game"}
      </Button>
    )
  }

  // Component that starts a solo game
  const SoloGameButton = ({classes}: any) => {
    const [requestCreateNewSoloGame, setRequestCreateNewSoloGame] = useState(false)
    const handleClick = useCallback(() => {
      setRequestCreateNewSoloGame(true)
    }, [requestCreateNewSoloGame])

    useEffect(() => {
      if (requestCreateNewSoloGame) {
        dispatch(createNewSoloGame())
      }
    }, [requestCreateNewSoloGame])

    return (
      <Button
        className={`${classes} mt-5`}
        variant="success"
        size="lg"
        onClick={!isCreatingSoloGame ? () => {handleClick()} : undefined}
        disabled={isSearchingForGame || isCreatingSoloGame}
      >
        {isCreatingSoloGame ? "Creating..." : "Solo Game"}
      </Button>
    )
  }

  return (
    <div>
      <div className="title mt-5">
        Welcome to the Trivia Quiz!
      </div>
      <div className="description mt-3">
        Compete with other players to see who can correctly answer the most questions.
      </div>
      <div className="description mt-3">
        Find a game below!
      </div>
      <div>
        <FindGameButton classes={"first-button"} />
        <SoloGameButton classes={""} />
      </div>
      <ul className="rules mt-5 mx-3 d-flex">
        <li>When the game begins, you will have 12 seconds to answer a multiple choice question.</li>
        <li>After submitting your answer the correct answer will be displayed in <span className="green">green</span>.</li>
        <li>Ties count as a loss (sorry!)</li>
      </ul>
    </div>
  )
}

export default Home