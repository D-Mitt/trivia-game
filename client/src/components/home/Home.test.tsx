import { mount } from "enzyme"
import * as React from "react"
import { Provider } from "react-redux"
import { createStore } from "redux"
import rootReducers from "../../store/index"
import Home from "./Home"

const createHome = () => {
  const store  = createStore(
    rootReducers
  )
  return (
    <Provider store={store}>
      <Home />
    </Provider>
  )
}

describe("Home", () => {
  const home = mount(createHome())

  test("matches the snapshot", () => {
    expect(home.html()).toMatchSnapshot()
  })

  test("descriptions/text are displayed", () => {
    const homeTitle = home.find(".title")
    expect(homeTitle).toBeTruthy()
    expect(homeTitle.at(0).text()).toEqual("Welcome to the Trivia Quiz!")

    const descriptions = home.find(".description")
    expect(descriptions.at(0).text()).toEqual("Compete with other players to see who can correctly answer the most questions.")
    expect(descriptions.at(1).text()).toEqual("Find a game below!")
  })

  test("buttons are displayed", () => {
    const findGameButton = home.find("FindGameButton").find("button")
    expect(findGameButton).toBeTruthy()
    expect(findGameButton.at(0).text()).toEqual("Find a Game")

    const soloGameButton = home.find("SoloGameButton").find("button")
    expect(soloGameButton).toBeTruthy()
    expect(soloGameButton.at(0).text()).toEqual("Solo Game")
  })

  test("list of rules displayed", () => {
    const listItems = home.find(".rules").find("li")
    expect(listItems.get(0)).toBeTruthy()
    expect(listItems.at(0).text()).toEqual("When the game begins, you will have 12 seconds to answer a multiple choice question.")

    expect(listItems.get(1)).toBeTruthy()
    expect(listItems.at(1).text()).toEqual("After submitting your answer the correct answer will be displayed in green.")

    expect(listItems.get(2)).toBeTruthy()
    expect(listItems.at(2).text()).toEqual("Ties count as a loss (sorry!)")
  })
})