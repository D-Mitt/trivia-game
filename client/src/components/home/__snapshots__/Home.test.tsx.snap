// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches the snapshot 1`] = `"<div><div class=\\"title mt-5\\">Welcome to the Trivia Quiz!</div><div class=\\"description mt-3\\">Compete with other players to see who can correctly answer the most questions.</div><div class=\\"description mt-3\\">Find a game below!</div><div><button type=\\"button\\" class=\\"first-button mt-5 btn btn-success btn-lg\\">Find a Game</button><button type=\\"button\\" class=\\" mt-5 btn btn-success btn-lg\\">Solo Game</button></div><ul class=\\"rules mt-5 mx-3 d-flex\\"><li>When the game begins, you will have 12 seconds to answer a multiple choice question.</li><li>After submitting your answer the correct answer will be displayed in <span class=\\"green\\">green</span>.</li></ul></div>"`;
